### PR TITLE
feat: use typescript-eslint@v6's reworked configs

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## Changes

Posting this as a reference for https://discord.com/channels/830184174198718474/845430950191038464/1120563248782114986:

> 👋 I'm working on testing out reworked preset configs for the next major version of typescript-eslint: https://github.com/typescript-eslint/typescript-eslint/issues/6760. I'd like to try them out on Astro because it's one of my favorite open source projects ❤️. Should I file an issue for this? (I could only find the bug template).
> Benefits of trying this out means I'll help out with the ESLint config in `withastro/astro` & listen even more closely & seriously to your feedback on the new preset configs!

_Edit: the response there was positive, so un-drafting! Thanks!_

Upgraded typescript-eslint to v6, with reworked ESLint configurations.

You can read https://typescript-eslint.io/blog/announcing-typescript-eslint-v6-beta#user-facing-breaking-changes for the rationale behind the config changes. Essentially, the new recommended starter configs are:

* `"plugin:@typescript-eslint/recommended-type-checked"`
* `"plugin:@typescript-eslint/stylistic-type-checked"`

I've commented any changes I suspect you might disagree with inline, for discussion. Would love to know what you all think!

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -1,23 +1,55 @@
...
+    '@typescript-eslint/no-unused-vars': ['error', { ignoreRestSiblings: true }],
+    'no-only-tests/no-only-tests': 'error',
+    '@typescript-eslint/no-shadow': ['error'],
+    'no-console': 'warn',
+
+    /*
+    interface: 404 results in 198 files
+    type: 129 results in 73 files
+    */
+    '@typescript-eslint/consistent-type-definitions': 'off',
```
Comment (JoshuaKGoldberg) [inline .eslintrc.cjs:24]: It looks like you've _somewhat_ solidified on preferring `interface` over `type`, but both have a good amount of usage. Is using one over the other when possible something you'd want to standardize?

https://typescript-eslint.io/rules/consistent-type-definitions

File: .eslintrc.cjs [hunk 2]
```diff
@@ -1,23 +1,55 @@
...
+    '@typescript-eslint/no-unsafe-call': 'off',
+    '@typescript-eslint/no-unsafe-member-access': 'off',
+    '@typescript-eslint/no-unsafe-return': 'off',
+    '@typescript-eslint/prefer-nullish-coalescing': 'off',
+    '@typescript-eslint/prefer-string-starts-ends-with': 'off',
+    '@typescript-eslint/require-await': 'off',
+    '@typescript-eslint/restrict-plus-operands': 'off',
+    '@typescript-eslint/restrict-template-expressions': 'off',
+    '@typescript-eslint/sort-type-constituents': 'off',
+    '@typescript-eslint/unbound-method': 'off',
```
Comment (JoshuaKGoldberg) [inline .eslintrc.cjs:49]: Each of these had at least a few existing warnings, so I didn't want to overreach and solve for them without asking. Are any of these rules ones you feel particularly strongly about?

- https://typescript-eslint.io/await-thenable
- https://typescript-eslint.io/ban-ts-comment
- https://typescript-eslint.io/class-literal-property-style
- https://typescript-eslint.io/consistent-indexed-object-style
- https://typescript-eslint.io/dot-notation
- https://typescript-eslint.io/no-base-to-string
- https://typescript-eslint.io/no-confusing-void-expression
- https://typescript-eslint.io/no-empty-function
- https://typescript-eslint.io/no-floating-promises
- https://typescript-eslint.io/no-misused-promises
- https://typescript-eslint.io/no-this-alias
- https://typescript-eslint.io/no-unsafe-argument
- https://typescript-eslint.io/no-unsafe-assignment
- https://typescript-eslint.io/no-unsafe-call
- https://typescript-eslint.io/no-unsafe-member-access
- https://typescript-eslint.io/no-unsafe-return
- https://typescript-eslint.io/prefer-nullish-coalescing
- https://typescript-eslint.io/prefer-string-starts-ends-with
- https://typescript-eslint.io/require-await
- https://typescript-eslint.io/restrict-plus-operands
- https://typescript-eslint.io/restrict-template-expressions
- https://typescript-eslint.io/sort-type-constituents
- https://typescript-eslint.io/unbound-method

If not, I can go ahead and fix for all their complaints.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -49,6 +49,9 @@ jobs:
       - name: Install dependencies
         run: pnpm install
 
+      - name: Build (ignoring failures)
+        run: pnpm run build || true
```
Comment (JoshuaKGoldberg) [inline .github/workflows/ci.yml:53]: Explaining: some newly enabled lint rules require type information. I'm under the impression `types` won't be fully filled out (i.e. some packages are seen as `any` / not-existing in editors) until `pnpm run build` happens. Is that the case?

File: package.json [hunk 4]
```diff
@@ -28,7 +28,7 @@
     "test:e2e": "cd packages/astro && pnpm playwright install && pnpm run test:e2e",
     "test:e2e:match": "cd packages/astro && pnpm playwright install && pnpm run test:e2e:match",
     "benchmark": "astro-benchmark",
-    "lint": "eslint --cache .",
+    "lint": "eslint . --report-unused-disable-directives",
```
Comment (JoshuaKGoldberg) [inline package.json:31]: Not strictly necessary, but I've found this to be a nice cleanup.

File: packages/astro/src/@types/astro.ts [hunk 5]
```diff
@@ -1087,9 +1087,13 @@ export interface AstroUserConfig {
...
-	integrations?: Array<
-		AstroIntegration | (AstroIntegration | false | undefined | null)[] | false | undefined | null
-	>;
+	integrations?: (
+		| AstroIntegration
+		| (AstroIntegration | false | undefined | null)[]
+		| false
+		| undefined
+		| null
+	)[];
```
Comment (JoshuaKGoldberg) [inline packages/astro/src/@types/astro.ts:1096]: Switching from `Array<...>` to `(...)[]` is very much a _`stylistic`_ choice (https://typescript-eslint.io/rules/array-type/). It seemed to me like `[]` was generally preferred already?

File: packages/markdown/remark/src/index.ts [hunk 6]
```diff
@@ -160,9 +160,8 @@ function prefixError(err: any, prefix: string) {
 	const wrappedError = new Error(`${prefix}${err ? `: ${err}` : ''}`);
 	try {
 		wrappedError.stack = err.stack;
-		// @ts-expect-error
 		wrappedError.cause = err;
```
Comment (JoshuaKGoldberg) [inline packages/markdown/remark/src/index.ts:163]: I added `target: ES2022` in the `tsconfig.json` so this line wouldn't need a `// @ts-expect-error`.

File: packages/astro/src/core/create-vite.ts [hunk 7]
```diff
@@ -32,7 +32,7 @@ import { joinPaths } from './path.js';
 interface CreateViteOptions {
 	settings: AstroSettings;
 	logging: LogOptions;
-	mode: 'dev' | 'build' | string;
+	mode: string;
```
Comment (bluwy) [inline packages/astro/src/core/create-vite.ts:35]: Are we able to keep the previous type definition? It _is_ redundant for now since the `string` would widen the type, but we have discussion to circle back in refactoring `mode` in the future, and keeping `'dev' | 'build'` keeps the history here.

File: package.json (hunk 4 above)
Comment (bluwy) [reply package.json:31]: I think we'd want to still keep the `--cache` too?

File: packages/astro/src/@types/astro.ts (hunk 5 above)
Comment (bluwy) [reply packages/astro/src/@types/astro.ts:1096]: I usually use `(...)[]` so this is fine by me.

File: .eslintrc.cjs (hunk 1 above)
Comment (bluwy) [reply .eslintrc.cjs:24]: I usually prefer `interface` over `type`. I think some of us prefer the other way around too. Maybe we can enforce this in a later PR, or if we decide on a style here.

File: packages/telemetry/src/system-info.ts [hunk 8]
```diff
@@ -52,7 +52,8 @@ export function getSystemInfo(versions: { viteVersion: string; astroVersion: str
 	}
 
 	const cpus = os.cpus() || [];
-	meta = {
+
+	return {
```
Comment (bluwy) [inline packages/telemetry/src/system-info.ts:56]: I think there's a bug here. We want to assign to `meta` so that it is cached on the second run. The new code doesn't assign to it anymore

File: packages/integrations/image/src/utils/workerPool.ts [hunk 9]
```diff
@@ -112,7 +112,7 @@ export default class WorkerPool<I, O> {
 		parentPort!.on('message', async (data) => {
 			const { msg, id } = data;
 			try {
-				const result = await cb(msg);
+				const result = cb(msg);
```
Comment (bluwy) [inline packages/integrations/image/src/utils/workerPool.ts:115]: I think we need an `await` here. It's usage at

https://github.com/joshuakgoldberg/astro/blob/e321967a408764407e0e4edd659d955a5f5db054/packages/integrations/image/src/vendor/squoosh/image-pool.ts#L148

returns a promise.

File: packages/integrations/image/src/index.ts [hunk 10]
```diff
@@ -39,7 +39,7 @@ export interface IntegrationOptions {
 	/**
 	 * Entry point for the @type {HostedImageService} or @type {LocalImageService} to be used.
 	 */
-	serviceEntryPoint?: '@astrojs/image/squoosh' | '@astrojs/image/sharp' | string;
+	serviceEntryPoint?: string;
```
Comment (bluwy) [inline packages/integrations/image/src/index.ts:42]: I think we'd want to preserve the previous static strings too 🤔

File: .github/workflows/ci.yml (hunk 3 above)
Comment (bluwy) [reply .github/workflows/ci.yml:53]: Yeah `pnpm run build` is needed to generate the types. I'd be fine with this since our CI also builds packages. I think this can be refactored further later to share the turbo cache in a later PR.

File: .eslintrc.cjs (hunk 1 above)
Comment (ematipico) [reply .eslintrc.cjs:24]: Usually, I go the other way around. 

`type` by default, and `interface` if I need to use its features:
- lazy loading
- `extends`


I don't have strong opinions, though.

File: .eslintrc.cjs (hunk 2 above)
Comment (ematipico) [reply .eslintrc.cjs:49]: > [typescript-eslint.io/await-thenable](https://typescript-eslint.io/await-thenable)

I'll keep it turned off. We have cases where we don't want to `await` a promise on purpose (middleware is one).

> [typescript-eslint.io/no-empty-function](https://typescript-eslint.io/no-empty-function)

Maybe we have a few no-ops in the code base? Not sure. I'd like to add this rule if possible and fix the cases where we trigger the error.

File: packages/astro/src/assets/vendor/image-size/types/pnm.ts [hunk 11]
```diff
@@ -19,7 +19,7 @@ const handlers: { [type: string]: Handler} = {
     let dimensions: string[] = []
 
     while (lines.length > 0) {
-      const line = lines.shift() as string
+      const line = lines.shift()!
```
Comment (ematipico) [inline packages/astro/src/assets/vendor/image-size/types/pnm.ts:22]: The bang operator is unsafe; I have a practice of leaving a `// SAFETY: <explanation>` comment where I explain why it's safe to do these kinds of casting.

**Review by ematipico — COMMENTED** [review]
I noticed that some changes use the bang operator `!`. I consider this very unsafe, and I wonder if we should change our code there or at least leave some comments explaining why it's safe. cc @withastro/maintainers-core

File: packages/astro/src/assets/vendor/image-size/types/pnm.ts (hunk 11 above)
Comment (ematipico) [reply packages/astro/src/assets/vendor/image-size/types/pnm.ts:22]: Yeah, both approaches require attention. I believe this is a chore from our side

File: .eslintrc.cjs (hunk 1 above)
Comment (bholmesdev) [reply .eslintrc.cjs:24]: I'm with Ema on that too. `type` is more ergonomic to write in my opinion, and I like that it avoids interface merging. Loose opinion strongly held :)

File: .eslintrc.cjs (hunk 2 above)
Comment (bholmesdev) [reply .eslintrc.cjs:49]: > https://typescript-eslint.io/rules/ban-ts-comment

This will cause issues when we reference experimental packages without generated types ([see here](https://github.com/withastro/astro/blob/main/packages/integrations/markdoc/src/index.ts#L16-L17)). I think these comments are rare, but have use cases we talk about during PR reviews. I'd prefer keeping this turned off!

File: packages/integrations/image/src/index.ts (hunk 10 above)
Comment (bholmesdev) [reply packages/integrations/image/src/index.ts:42]: `string` will override those static strings in intellisense, so I understand the deletion. I actually _just_ went through an extensive twitter thread on this 😄 

TL;DR: if you want to show suggestions via intellisense, but _still_ allow arbitrary strings, you can [either append `& {}` to the definition or use the uppercase `String` object instead](https://twitter.com/BHolmesDev/status/1670575170027769856):

`serviceEntryPoint?: '@astrojs/image/squoosh' | '@astrojs/image/sharp' | String;`

File: .eslintrc.cjs (hunk 1 above)
Comment (bluwy) [reply .eslintrc.cjs:24]: I don't have a strong reason picking `interface` over `type`, other than I use them the same way as `function foo() {}` vs `const foo = () => {}`. E.g. `interface` for typing object shapes, `type` for quick one-offs. But I also somewhat like that `interface` generally provides nicer TS intellisense message, especially when unions/extends are used.

Comment (matthewp) [reply .eslintrc.cjs:24]: I think we should avoid doing an upgrade and changing (or adding) lint rules in the same PR. Those are two separate ideas that should be discussed separately.

File: .eslintrc.cjs (hunk 2 above)
Comment (matthewp) [reply .eslintrc.cjs:49]: Same comment as above, would prefer that we don't change the rules in an upgrade PR.

File: packages/astro/src/@types/astro.ts (hunk 5 above)
Comment (lilnasy) [reply packages/astro/src/@types/astro.ts:1096]: I'd argue against linting for either. `(...)[]` looks neat with smaller types but becomes easy to miss with larger unions and object types.

File: .eslintrc.cjs (hunk 1 above)
Comment (JoshuaKGoldberg) [reply .eslintrc.cjs:24]: I'll leave this as a todo then - agreed that if there isn't team consensus, this is not a great place to come to one. Thanks!

@bholmesdev looking forward to a whiteboard video on the nuanced differences between the two 😉

File: package.json (hunk 4 above)
Comment (JoshuaKGoldberg) [reply package.json:31]: Ehh I've found the cache to be ... unreliable when working with type checking. Theoretically a file's lint result could be impacted by a couple of other factors -other files it imports types from, TSConfig settings- so caching becomes difficult.

File: packages/astro/src/@types/astro.ts (hunk 5 above)
Comment (JoshuaKGoldberg) [reply packages/astro/src/@types/astro.ts:1096]: @lilnasy does https://typescript-eslint.io/rules/array-type/#array-simple match your preference?

File: packages/telemetry/src/system-info.ts (hunk 8 above)
Comment (JoshuaKGoldberg) [reply packages/telemetry/src/system-info.ts:56]: Ah, this is what I get for hand-editing 😅 good catch. Will undo.

File: packages/astro/src/core/create-vite.ts (hunk 7 above)
Comment (JoshuaKGoldberg) [reply packages/astro/src/core/create-vite.ts:35]: Yup we can! I can see why you'd want this. I'll move https://typescript-eslint.io/rules/no-redundant-type-constituents to the _todo_ pile.

File: packages/integrations/image/src/utils/workerPool.ts (hunk 9 above)
Comment (JoshuaKGoldberg) [reply packages/integrations/image/src/utils/workerPool.ts:115]: I'd initially had https://typescript-eslint.io/rules/require-await on the enabled list & fixed up its complaints, then realized you use `await` for places that _might_ give a Promise from middleware (as stated elsewhere in the review). I missed reverting this change. Will revert this and anything else similar I can spot.

File: packages/integrations/image/src/index.ts (hunk 10 above)
Comment (JoshuaKGoldberg) [reply packages/integrations/image/src/index.ts:42]: As much as I love working around TypeScript 😄 I'll revert these string changes & move https://typescript-eslint.io/rules/no-redundant-type-constituents/ to the todo pile.

File: .eslintrc.cjs (hunk 2 above)
Comment (JoshuaKGoldberg) [reply .eslintrc.cjs:49]: > > [typescript-eslint.io/rules/ban-ts-comment](https://typescript-eslint.io/rules/ban-ts-comment)
> 
> This will cause issues when we reference experimental packages without generated types ([see here](https://github.com/withastro/astro/blob/main/packages/integrations/markdoc/src/index.ts?rgh-link-date=2023-06-20T11%3A05%3A25Z#L16-L17)). I think these comments are rare, but have use cases we talk about during PR reviews. I'd prefer keeping this turned off!

Maybe helpful?: it can be configured to allow the comment directives _if_ there's a description attached.

> Same comment as above, would prefer that we don't change the rules in an upgrade PR.

👍 I'll keep these all as _todo_.

File: packages/astro/src/assets/vendor/image-size/types/pnm.ts (hunk 11 above)
Comment (JoshuaKGoldberg) [reply packages/astro/src/assets/vendor/image-size/types/pnm.ts:22]: Yup - `!` is precisely the same as `as string` when used on a nullish `string`. All this change is doing is going with a slightly more succinct syntax. https://typescript-eslint.io/rules/non-nullable-type-assertion-style

File: packages/astro/src/@types/astro.ts (hunk 5 above)
Comment (lilnasy) [reply packages/astro/src/@types/astro.ts]: That's perfect!

File: packages/astro/src/assets/vendor/image-size/types/pnm.ts (hunk 11 above)
Comment (ematipico) [reply packages/astro/src/assets/vendor/image-size/types/pnm.ts:22]: Thank you. I suppose we can keep the existing changes, then we can enable the lint rule and then we can refactor the errors :) 

The bang operator is easily detectable by an analyzer

File: packages/astro/src/@types/astro.ts (hunk 5 above)
Comment (bluwy) [reply packages/astro/src/@types/astro.ts]: I'm not really sure about `array-simple`. If we're undecided about this, I think we can also leave it unchanged for now (sorry for causing more changes)

File: package.json (hunk 4 above)
Comment (bluwy) [reply package.json:31]: I won't deny that I've seen `--cache` acting strange before, so I could be persuaded removing it for now if there's no rejections from others.

File: .eslintrc.cjs (hunk 1 above)
Comment (nickmccurdy) [reply .eslintrc.cjs]: Note that changing from `interface` to `type` can be a breaking change, both due to inability to merge declarations and potential performance regressions.

File: packages/astro/src/core/create-vite.ts (hunk 7 above)
Comment (nickmccurdy) [reply packages/astro/src/core/create-vite.ts]: Note that allowing redundant string literals does still improve autocomplete.

File: packages/astro/src/assets/vendor/image-size/types/pnm.ts (hunk 11 above)
Comment (nickmccurdy) [reply packages/astro/src/assets/vendor/image-size/types/pnm.ts:22]: Hi, I wrote the original PR to Vite replacing `as` with `!`. The reason for this is that while both can make potentially unsafe assertions, `as` can silence type errors when types are assignable, making `!` relatively more safe.

File: packages/astro/src/core/create-vite.ts (hunk 7 above)
Comment (bholmesdev) [reply packages/astro/src/core/create-vite.ts]: @nickmccurdy Hm, doesn't seem to from my experience. Union-ing with a lowercase `string` will blow away other string literals in VS Code. There's a few hacks you can use to get around this like adding `& {}` or `String` to the type.

File: .eslintrc.cjs (hunk 1 above)
Comment (JoshuaKGoldberg) [reply .eslintrc.cjs]: FWIW at this point `type` has been optimized pretty significantly. I can't remember the last time I've seen it be a performance concern in any code (and would love to be pointed to a demonstrable example!). The TS wiki page is out of date.

File: packages/astro/src/@types/astro.ts (hunk 5 above)
Comment (JoshuaKGoldberg) [reply packages/astro/src/@types/astro.ts]: No worries 😄 I'll go through and undo the changes, then merge from `main` & re-request your review. Thanks!

Comment (JoshuaKGoldberg) [reply packages/astro/src/@types/astro.ts]: Done. This is a lesson learned the hard way - when there are many complaint from a stylistic rule, leave it out in the proposal. 😄

**Comment by JoshuaKGoldberg** [comment]
> **CI / Lint (pull_request)** Cancelled after 3m

...works on my machine! (as in, `pnpm lint` after `pnpm build` produces no errors. Undrafting 🤷

**Review by bluwy — APPROVED** [review]

File: .eslintrc.cjs [hunk 12]
```diff
@@ -1,23 +1,53 @@
 module.exports = {
+  extends: [
```
Comment (ematipico) [inline .eslintrc.cjs:2]: I wonder, since this is a breaking change, should we try to move to use the new eslint configuration format? https://eslint.org/docs/latest/use/configure/configuration-files-new

**Review by ematipico — APPROVED** [review]
